### DISCUSSION TITLE
Program Publish API "On After Publish callback failed" error #46

### DISCUSSION THREAD

**Discussion by githubpiyush** [Sunbird-coKreat/Community#39] [General Discussions]
We are trying to publish a program using API in VDN staging env and getting this error :-

API :- **https://dockstaging.sunbirded.org/api/program/v1/publish**

`{ "id": "api.v1.publish", "ver": "1.0", "ts": "2022-12-06T11:25:25.928Z", "params": { "resmsgid": "ae3eda80-7558-11ed-bcf5-df48f63ad78b", "msgid": null, "status": "failed", "err": "PRG_PRPUB" }, "responseCode": "PRG_PRPUB02_ERR_PUBLISH_PROGRAM", "result": "On After Publish callback failed" }
`

Please let us know how can we resolve this error.

**Comment by Nadiya496** [comment]
Hi @githubpiyush ,

Successfully able to publish the Project. Please share the request payload for further checking.
cc" @pallakartheekreddy

**Reply by githubpiyush** [reply]
@Nadiya496 
Can you try this 3 curls and check whether the publish is working or not?

1. API :- **https://dockstaging.sunbirded.org/api/program/v1/create**
    Request body :- `{
    "request": {
        "name": "MIGRATED Survey of Diksha awareness – Test(Support testing) sourcing project",
        "description": "Survey of Diksha awareness – Test(Support testing) sourcing project description",
        "nomination_enddate": "2022-12-07T11:17:07.466Z",
        "rewards": null,
        "shortlisting_enddate": "2022-12-07T11:17:07.466Z",
        "enddate": "2022-12-08T11:17:07.466Z",
        "content_submission_enddate": "2022-12-07T11:17:07.466Z",
        "type": "public",
        "target_type": "searchCriteria",
        "content_types": [],
        "target_collection_category": [],
        "sourcing_org_name": "Staging Custodian Organization",
        "rootorg_id": "0126796199493140480",
        "createdby": "7ff56e27-4eec-475e-bb77-8457e34244a2",
        "createdOn": "2022-12-06T11:17:07.466Z",
        "startdate": "2022-12-07T11:17:07.466Z",
        "slug": "sunbird",
        "status": "Draft",
        "program_id": "",
        "rolemapping": [],
        "config": {
            "defaultContributeOrgReview": false,
            "roles": [
                {
                    "id": 1,
                    "name": "CONTRIBUTOR",
                    "tabs": [
                        1
                    ],
                    "default": true,
                    "defaultTab": 1
                },
                {
                    "id": 2,
                    "name": "REVIEWER",
                    "tabs": [
                        2
                    ],
                    "defaultTab": 2
                }
            ]
        }
    }
}`

2. API :- **https://dockstaging.sunbirded.org/api/program/v1/update**
   Request body :-
   `{
    "request": {
        "name": "MIGRATED !! Survey of Diksha awareness – Test(Support testing) sourcing project",
        "config": {
            "defaultContributeOrgReview": false,
            "roles": [
                {
                    "id": 1,
                    "name": "CONTRIBUTOR",
                    "tabs": [
                        1
                    ],
                    "default": true,
                    "defaultTab": 1
                },
                {
                    "id": 2,
                    "name": "REVIEWER",
                    "tabs": [
                        2
                    ],
                    "defaultTab": 2
                }
            ],
            "framework": [
                "ekstep_ncert_k-12"
            ],
            "frameworkObj": {
                "code": "ekstep_ncert_k-12",
                "name": "CBSE",
                "type": "K-12",
                "identifier": "ekstep_ncert_k-12"
            },
            "sharedContext": []
        },
        "targetprimarycategories": [
            {
                "identifier": "obj-cat:observation_questionset_all",
                "name": "Observation",
                "targetObjectType": "QuestionSet"
            },
            {
                "identifier": "obj-cat:survey_questionset_all",
                "name": "Survey",
                "targetObjectType": "QuestionSet"
            }
        ],
        "targetprimarycategorynames": [
            "Observation",
            "Survey"
        ],
        "program_id": "855ea3d0-7557-11ed-bcf5-df48f63ad78b"
    }
}`

3. API :- **https://dockstaging.sunbirded.org/api/program/v1/publish**
   Request body :-
   `{
    "request": {
       "program_id": "855ea3d0-7557-11ed-bcf5-df48f63ad78b",
        "channel": "sunbird"
    }
}`

**Reply by githubpiyush** [reply]
@Nadiya496 
Please update here once it is resolved.

CC :-  @aks30

**Comment by Nadiya496** [comment]
@githubpiyush Please share the user credentials of the ORG_ADMIN.

**Reply by githubpiyush** [reply]
@Nadiya496 
shared in slack channel, let me know if you are able to login or not.

**Comment by vaivk369** [comment]
The user with ORG_ADMIN role of the state organizations can only create and publish the project. The organization used as a rootorg_id in the API provided in the create API is a custodian org. Please correct and provide state organization id and created_by should be the user id of the ORG_ADMIN of the same organization.
